### PR TITLE
Don't close connection on socket error + fix tests to newer driver behavior

### PR DESCRIPTION
**Pull request by ashelley**
Also add test commands to package.json so it's easier to discover how to run unit tests

### CODE REVIEW

**Comment by ashelley** [comment]
Other than my horrible commit message one issue I don't think is very good is that the error emitted on the connection pool is of type ConnectionError as per:

https://github.com/pekim/tedious/blob/e3f0a59f213e97bf284b32922aea0969af537a52/src/connection.js#L485

but the error that gets emitted to the request isn't of the same type:

https://github.com/pekim/tedious/blob/e3f0a59f213e97bf284b32922aea0969af537a52/src/connection.js#L487

However this above nuance might be the expected result.  

Also note I changed the behaviour of the 'lost connection handling' test to expect that the request callback gets called.

**Comment by ben-page** [comment]
Hi, thanks for this pull request.

At the moment, I'm not convinced this is the correct solution. It seems to me that it's good practice to call `connection.close` in the event of an error. I agree that this should not stop tedious from calling the request callback, though. I'd prefer to see this fixed in tedious.

**Comment by ben-page** [comment]
I'm keeping this open for now in case I'm persuaded otherwise.

File: lib/connection-pool.js [hunk 1]
```diff
@@ -82,15 +82,15 @@ function createConnection(pooled) {
         self.log('connection closing because of error');
 
         connection.removeListener('end', endHandler);
-        connection.close();
-
+        if(err.code !== 'ESOCKET') {
+            connection.close();
```
Comment (arthurschreiber) [inline lib/connection-pool.js]: You never need to close the socket, no matter what error gets emitted.

**Comment by ashelley** [comment]
Regardless of how the 'request callback problem' gets fixed I do have a fix for tests in this patch so that the tests pass with the change found here:

https://github.com/pekim/tedious/pull/330

So we'll want the test fixes part of this patch once a new tedious release is pushed out.

**Comment by ben-page** [comment]
I see understand the issue better now. If you please remove err.code test and the call to `connection.close` entirely, I'll pull this PR.

File: test/connection-pool.test.js [hunk 2]
```diff
@@ -1,16 +1,13 @@
...
-var connectionConfig = {
-    userName: 'test',
-    password: 'test',
-    server: 'dev1',
-    options: {
-        appName: 'pool-test',
-        database: 'test'
-    }
-};
+function getConfig() {
```
Comment (ben-page) [inline test/connection-pool.test.js]: What's the reasoning for this? I seems like having the connection config in an external file that is not included in the project is unnecessarily convoluted.

Comment (ashelley) [inline test/connection-pool.test.js]: I actually copied this logic from tedious.  I was trying to test against my sql server and needed different connection settings to test.  The main reason was that I added a port option.  I probably could have modified my hosts file and sql server to facilitate running with the existing test config.

https://github.com/pekim/tedious/blob/master/test/integration/connection-test.coffee

I wasn't sure the best way to facilitate this.

**Comment by ashelley** [comment]
I've modified the PR to remove the test for err.code and remove the call to connection.close()

**Comment by ben-page** [comment]
It looks to me that the tests you added are really testing Tedious rather than the connection pool. Am I missing the point?

**Comment by ashelley** [comment]
They were there to test the interaction between the pool and the driver.  Before the patch these tests would fail so that's why I added them.

**Comment by ben-page** [comment]
Ok. Well I removed them. The changes you made to the test "lost connection handling" (which I renamed "connection error handling") tested for the problem. We don't need three tests for the same issue.

**Comment by ben-page** [comment]
I'm closing this PR, but I created another PR based on this patch.

**Comment by ashelley** [comment]
Sounds good by me... I just wanted to point out that the "connection error handling" test covers a different code path as it is invoked via "requestTimeout" timer which explicitly called via a setTimeout and the request callback is explicitly called.

https://github.com/pekim/tedious/blob/master/src/connection.js#L1094 

The newly added tests were to ensure that when a "connection error" occurs:

https://github.com/pekim/tedious/blob/master/src/connection.js#L483

The request callback is still called which is what the new code fixes.  

Now that the pool doesn't care to close the connection on these types of errors it probably doesn't matter to test this at the pool level I just wanted to note that the test was testing something other than what the "connection error handling" was testing.

Anyways thank you for taking the time for getting this problem fixed up!

**Pull request closed without merging**